Perform cross-contract taint analysis from diff of two upgrade versions

webthethird opened the pull request:
In `slither.utils.upgradeability`, the `compare` function found functions and variables in the contracts being compared that were tainted due to new or modified functions. But functions and variables in other contracts may also be tainted by these changes due to external calls.

This PR currently adds two functions to `slither.utils.upgradeability`, one of which is used by `compare` to perform cross-contract taint analysis and return more information in the diff results:

- `tainted_external_contracts` takes a list of functions, i.e., the new, modified and tainted functions found in `compare`, and searches them for high-level calls to other contracts. It marks the corresponding functions (and public variables) as tainted, as well as any variables they write to. Then, it searches for any other functions in the external contract which read or write to tainted variables. It returns a list of `TaintedExternalContract` objects, which are typed dicts containing the contract object plus two lists of tainted functions and variables. `compare` calls this function, after finding all new, modified and tainted functions in the diffed contracts, and returns the list along with all original outputs. 
- `tainted_inheriting_contracts` addresses a shortcoming of `tainted_external_contracts`, which is that high-level calls may take advantage of polymorphism and call functions from a base contract, causing the method to miss tainted functions and variables defined in contracts that inherit the base class. `tainted_inheriting_contracts` takes a list of `TaintedExternalContract` objects as input, as well as an optional list of contracts to search for inheritance, and outputs a potentially augmented list of tainted contracts with their tainted functions and variables. It searches the list of contracts (or the contracts in the compilation unit) for contracts that inherit from tainted contracts, and searches their declared functions for internal calls to previously tainted functions or reads/writes to tainted variables. It then appends the new `TaintedExternalContract` to the original list.

These functions are not strictly related to upgradeability, so I'm not sure if they belong here or somewhere else. I am open to any suggestions/feedback!

montyly commented on slither/utils/upgradeability.py:
```diff
@@ -159,16 +174,137 @@ def compare(
+            for f in c.functions_declared:
+                internal_calls = f.all_internal_calls()
+                if any(
+                    str(call) == str(t) for t in tainted["functions"] for call in internal_calls
```
I think comparing a string here might lead to false result, as `str` will just return the name. Maybe we can compare the `canonical_name` instead?

webthethird commented on slither/utils/upgradeability.py:
```diff
@@ -61,11 +64,23 @@
 from slither.tools.read_storage.read_storage import SlotInfo, SlitherReadStorage
 
 
+class TaintedExternalContract(TypedDict):
```
I went ahead and made `TaintedExternalContract` a normal class. 

I also added `TaintedFunction` and `TaintedVariable` classes, each of which have a `tainted_by: List[Union[Function, StateVariable]]` property to keep track of which function/variable tainted which. Do you think that's overkill? I could just go back to having `TaintedExternalContract` store a list of Functions and a list of StateVariables, rather than TaintedFunctions and TaintedVariables.

webthethird replied:
I reverted the addition of `TaintedVariable` and `TaintedFunction`, since those mostly just complicated the code. I still have the `TaintedExternalContract` as a normal class, which stores a list of `Function`s and `StateVariable`s.

webthethird commented on slither/utils/upgradeability.py:
```diff
@@ -273,7 +409,7 @@ def encode_ir_for_compare(ir: Operation) -> str:
         return f"({encode_var_for_compare(ir.lvalue)}):=({encode_var_for_compare(ir.rvalue)})"
     if isinstance(ir, Index):
-        return f"index({ntype(ir.index_type)})"
+        return f"index({ntype(ir.variable_right.type)})"
```
It only became wrong when the `index_type` property was removed from the `Index` class. Slither was just crashing when I tried to access the property, since it didn't exist anymore. Is that still something we can test?

webthethird commented:
>I think we need to add tests, and inline code documentation to explain the different steps of the new functions

I've updated the test in test_upgradeability_util, adding some external taint to the test contracts, and I added some in-line comments.

webthethird commented:
I believe this one is ready for another review @montyly or @0xalpharush
